feat: add iOS target

FunkyMuse opened the pull request:
https://github.com/frostwire/frostwire-jlibtorrent/issues/287

I've ran into some issues, if I run 

```./build-ios-arm64.sh```

Then for some reason the `src/main/java/com.frostwire.jlibtorrent.swig` folder gets deleted and the gradle task is unable to finish the build because of missing imports since the swigs are deleted.

I'll need some help @gubatron with this if possible, i've been hitting a dead end for an hour already, I hope my head clears tomorrow so I'd fix this without any assits

FunkyMuse commented:
hmm, it seems that no matter which script i run, it always deletes those files, strange

FunkyMuse commented:
<img width="757" alt="Screenshot 2023-12-18 at 23 52 55" src="https://github.com/frostwire/frostwire-jlibtorrent/assets/25645122/3b44bf35-3e49-4f08-a10a-a91cb507704c">

up until this point everything works good

FunkyMuse commented:
<img width="757" alt="Screenshot 2023-12-19 at 00 04 00" src="https://github.com/frostwire/frostwire-jlibtorrent/assets/25645122/adf83d3c-acc9-4447-bee8-9daa40648ff0">

after commenting out line 49 in run-swig.sh

build is successful but there's an error

<img width="1602" alt="Screenshot 2023-12-19 at 00 04 24" src="https://github.com/frostwire/frostwire-jlibtorrent/assets/25645122/b6106572-e4a6-4027-b076-f4d9c12caf39">

gubatron commented:
all the deleted code on that folder is generated by SWIG, this is necessary, otherwise your java code will not match the latest changes in the libtorrent library.

you're having some sort of C++ build error because you didn't get to have a .dylib created.

gubatron commented:
(A lot of changes and cleanup are coming on the `swig/` folder from [a branch](https://github.com/frostwire/frostwire-jlibtorrent/tree/2.0/swig) I hope to have ready the next spring, which wraps libtorrent 2.0)

FunkyMuse commented:
Since i shouldn't comment out that code that deletes the folder but without it, the script can't continue 🤔

At least am I on the right path, I've tried adapting the scripts as much as I can from what you already had in place

Feel free to review the code if you have time

FunkyMuse commented:
no matter what script i run, the build fails 
<img width="1289" alt="image" src="https://github.com/frostwire/frostwire-jlibtorrent/assets/25645122/3d9df9ca-f3c4-47ca-b877-34eaedfdf20b">

FunkyMuse commented:
```
...found 6 targets...
...updating 5 targets...
darwin.compile.c++ bin/release/ios/arm64/32.o
clang: warning: using sysroot for 'MacOSX' but targeting 'iPhone' [-Wincompatible-sysroot]
/Users/funkymuse/src/boost_1_82_0/libs/config/checks/architecture/32.cpp:9:10: error: 'test' declared as an array with a negative size
int test[sizeof(void*) == 4? 1 : -1];
         ^~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.

    "clang++"   -fvisibility-inlines-hidden -std=c++17 -miphoneos-version-min=14.0 -fno-strict-aliasing -fvisibility=hidden -D_DARWIN_C_SOURCE -fPIC -O0 -fno-inline -Wall -g -fvisibility=hidden -dynamic -gdwarf-2 -fexceptions  -DBOOST_ALL_NO_LIB=1 -DBOOST_NO_IOSTREAM -DTORRENT_USE_I2P=1 -D_WIN32_WINNT=0x0600  -I"/Users/funkymuse/src/boost_1_82_0"  -c -o "bin/release/ios/arm64/32.o" "/Users/funkymuse/src/boost_1_82_0/libs/config/checks/architecture/32.cpp"

...failed darwin.compile.c++ bin/release/ios/arm64/32.o...
...failed updating 1 target...
...updated 4 targets...
...found 2 targets...
...updating 1 target...
darwin.compile.c++ bin/release/ios/arm64/64.o
clang: warning: using sysroot for 'MacOSX' but targeting 'iPhone' [-Wincompatible-sysroot]
...updated 1 target...
...found 2 targets...
...updating 1 target...
darwin.compile.c++ bin/release/ios/arm64/arm.o
clang: warning: using sysroot for 'MacOSX' but targeting 'iPhone' [-Wincompatible-sysroot]
...updated 1 target...
...found 1 target...
...updating 1 target...
config-cache.write bin/project-cache.jam
...updated 1 target...

```

hmmm

FunkyMuse commented:
```
...skipped <pbin/release/ios/arm64>libjlibtorrent.so for lack of <pbin/release/ios/arm64>libtorrent_jni.o...
```
:(

FunkyMuse commented:
@gubatron i'd appreciate your help on this PR

gubatron commented:
wish I could be of help, currently dealing with much harder issues merging libtorrent 2 (while dealing with holiday visits). Once I'm done with the libtorrent 2 branch I can take a look, as there's several fixes to run-swig.sh and other scripts. (much better project structure now)

gubatron commented on build.gradle:
```diff
@@ -127,6 +136,7 @@ artifacts {
     archives sourcesJar
     archives javadocJar
     archives nativeMacOSArm64Jar
+    archives nativeiOSArm64Jar
```
I'd put this one last :)

gubatron commented on swig/build-utils.shinc:
```diff
@@ -47,6 +47,27 @@ macosx_env() {
+    export IOS_COMPILER="iphoneos"
+
+    XCODE_SDK_PATH=$(xcode-select -p)/Platforms/${IOS_PLATFORM}.platform/Developer/SDKs/${IOS_PLATFORM}${IOS_SDK_VERSION}.sdk
+    export IOS_SYSROOT="${XCODE_SDK_PATH}"
```
hmm

gubatron replied:
If you're not going to use XCODE_SDK_PATH elsewhere, why not just:
`export IOS_SYSROOT=$(xcrun --sdk macosx --show-sdk-path)`

gubatron replied:
no need for a XCODE_SDK_PATH variable, just do
```
export IOS_SYSROOT=$(xcrun --sdk iphoneos --show-sdk-path)
```

gubatron commented on swig/build-utils.shinc:
```diff
@@ -253,6 +274,12 @@ build_libraries() {
     mv swig/bin/release/${os_build}/${os_arch}/${SHARED_LIB} swig/bin/release/${os_build}/${os_arch}/${RELEASE_SHARED_LIB}
     cp swig/bin/release/${os_build}/${os_arch}/${RELEASE_SHARED_LIB} .
   fi
+  if [[ $os_build == "ios" ]]; then
```
please DRY (Don't repeat yourself), just add an OR condition on the "macosx" if, the code block you added is verbatim the same as the "maxosx" one.

gubatron commented on swig/config/ios-arm64-config.jam:
```diff
@@ -0,0 +1,10 @@
+    <cxxflags>-fno-strict-aliasing
+    <cxxflags>-fvisibility=hidden
+    <cxxflags>-D_DARWIN_C_SOURCE
+    <linkflags>"-isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk"
```
I would not hardcode paths here, use environment variables:

e.g.:
```
ANDROID_TOOLCHAIN = [ os.environ ANDROID_TOOLCHAIN ] ;
ANDROID_API = [ os.environ android_api ] ;

using clang-linux : arm : $(ANDROID_TOOLCHAIN)/bin/armv7a-linux-androideabi$(ANDROID_API)-clang++ :
```

I think you already define that path in `build-ios-arm64.sh`

once again DRY principle, otherwise when there's a new path you have to fix 2 files and waste time figuring out why shit doesn't work. Keep it simple.

gubatron replied:
![Screenshot 2023-12-29 at 7 57 55 PM](https://github.com/frostwire/frostwire-jlibtorrent/assets/163977/712bf97e-2d38-49a1-ae7d-83bfd795229e)

gubatron replied:
Thank you for this.
I'm sure we'll make this happen.
You are close.

gubatron commented on swig/config/ios-arm64-config.jam:
```diff
@@ -0,0 +1,10 @@
+
+using darwin : arm64 : clang :
+    <cxxflags>-std=c++17
+    <cxxflags>"-miphoneos-version-min=17.2"
```
you shouldn't need these quotes "" here

macos version:
```
using darwin : arm64 : clang++ :
    <cxxflags>-std=c++17
    <cxxflags>-mmacosx-version-min=11.3
    <cxxflags>-fno-strict-aliasing
    <cxxflags>-fvisibility=hidden
    <cxxflags>-D_DARWIN_C_SOURCE
    ;
```

gubatron commented:
```
maybe you're using an older version of Boost?
```
On my libtorrent 2 branch I'm using boost 1.82.0 and I don't have this issue, perhaps you want to check your build environment, your clang version.

This is mine:
```
 clang --version
Apple clang version 15.0.0 (clang-1500.0.40.1)
Target: arm64-apple-darwin23.2.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

Here, there's some stuff to pay attention
```
...found 6 targets...
...updating 5 targets...
darwin.compile.c++ bin/release/ios/arm64/32.o
clang: warning: using sysroot for 'MacOSX' but targeting 'iPhone' [-Wincompatible-sysroot]
/Users/funkymuse/src/boost_1_82_0/libs/config/checks/architecture/32.cpp:9:10: error: 'test' declared as an array with a negative size
int test[sizeof(void*) == 4? 1 : -1];
         ^~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

This one raises my eyebrow:
`clang: warning: using sysroot for 'MacOSX' but targeting 'iPhone' [-Wincompatible-sysroot]`

In my copy of boost, I have the same exact test and it's ok in my build environment.
![Screenshot 2023-12-29 at 8 03 49 PM](https://github.com/frostwire/frostwire-jlibtorrent/assets/163977/baf1c58f-374d-41a4-bbcb-d1d214a25426)

To solve issues like this, I basically comment out code in the build script, I'd make sure that boost gets built correctly.
by doing this on `swig/build-ios-arm64.sh`

```
#!/usr/bin/env bash
# This script is meant to run here inside the swig folder
# It's supposed to be a one-step build for the java jar and iOS (.dylib enclosing) jar
# Output .jar files will be at:
# ../build/libs/${LIBRARY_NAME}-<version>.jar
# ../build/libs/${LIBRARY_NAME}-ios-<version>.jar
source build-utils.shinc
ios_env
common_env
check_min_req_vars
export os_arch=arm64
export os_build=ios
export SHARED_LIB=lib${LIBRARY_NAME}.dylib
export RELEASE_SHARED_LIB=lib${LIBRARY_NAME}.${os_arch}.dylib
export CXX=clang++
export CC=clang
export run_openssl_configure="./Configure iphoneos-${os_arch} -fembed-bitcode ${OPENSSL_NO_OPTS} --prefix=${OPENSSL_ROOT}"
export run_readelf="otool -L bin/release/${os_build}/${os_arch}/${SHARED_LIB}"
export run_bjam="${BOOST_ROOT}/b2 -j16 --user-config=config/${os_build}-${os_arch}-config.jam variant=release toolset=darwin-iphone target-os=iphone location=bin/release/${os_build}/${os_arch}"
export run_objcopy="echo dummy run_objcopy for ${os_build} ${os_arch}"
export run_strip="strip -S -x bin/release/${os_build}/${os_arch}/${SHARED_LIB}"
export run_native_jar="./gradlew nativeiOSArm64Jar"
create_folder_if_it_doesnt_exist ${SRC}
prompt_msg "$0: About to prepare BOOST ${BOOST_VERSION}"
press_any_to_continue
prepare_boost #make the script work up to here and make sure it's building fine, passing tests and what not. Perhaps there's something different in the way that clang works for iOS and that C++ doesn't work, and maybe there's a flag you can pass to `run_bjam` so it doesn't perform that test.
#prepare_openssl
#build_openssl
#prepare_libtorrent
#./run-swig.sh
#build_libraries
```

gubatron commented:
In case you didn't catch the comment above:

`make the script work up to here and make sure it's building fine, passing tests and what not. Perhaps there's something different in the way that clang works for iOS and that C++ doesn't work, and maybe there's a flag you can pass to `run_bjam` so it doesn't perform that test.`

gubatron commented on build.gradle:
```diff
@@ -136,14 +136,14 @@ artifacts {
     archives nativeAndroidX86Jar
     archives nativeAndroidArm64Jar
     archives nativeAndroidX64Jar
+    archives nativeiOSArm64Jar
```
👍

gubatron commented on swig/build-utils.shinc:
```diff
@@ -268,18 +268,12 @@ build_libraries() {
     cp swig/bin/release/${os_build}/${os_arch}/${SHARED_LIB} .
   fi
-  if [[ $os_build == "macosx" ]]; then
+  if [[ $os_build == "macosx" || $os_build == "ios" ]]; then
```
👍

gubatron commented on swig/config/ios-arm64-config.jam:
```diff
@@ -1,10 +1,9 @@
 import os ;
 
-using darwin : arm64 : clang :
+using darwin : arm64 : clang++ :
```
ha!

gubatron commented:
Happy New Year, thanks for your work on this

FunkyMuse commented:
> Happy New Year, thanks for your work on this

Happy New Year to you too, hope you have a nice one.

gubatron commented:
maybe now it's time to restart this task, I've finally merged the 2.0 branch and a lot got cleaner, more modular, better.

FunkyMuse commented:
> maybe now it's time to restart this task, I've finally merged the 2.0 branch and a lot got cleaner, more modular, better.

Quite a lot of time has passed, i've lost context of what was going on, maybe start from scratch?
Or if you have better idea let me know, if you need help from me at all

gubatron commented:
I think I'll look at some of what you did here and once my build-macos-arm64.sh is 100% there, I'll clone it as build-ios-arm64.sh and tweak it

# CoPilot Changes Summary:

This pull request introduces a new build script for iOS on ARM64 architecture and updates the build utilities to support this new target. The most important changes include the addition of the iOS build script, the new environment configuration for iOS, and modifications to the build utilities to handle iOS builds.

### New iOS Build Script:

* [`swig/build-ios-arm64.sh`](diffhunk://#diff-13f667e1641bd79cc4e52d254a2f4b295db69dde11d186b69527b98125c5d2a1R1-R31): Added a new script to build the Java JAR and iOS (.dylib) JAR for ARM64 architecture. This script sets up the necessary environment and runs the build process.

### Environment Configuration for iOS:

* [`swig/build-utils.shinc`](diffhunk://#diff-b921a6dcfbd452736ae52fd0c2a7ec44f27291a93ad18565e33ccdf31101a394R50-R69): Introduced `ios_env()` function to set up the environment variables specific to iOS builds, including paths and compiler flags.

### Build Utilities Update:

* [`swig/build-utils.shinc`](diffhunk://#diff-b921a6dcfbd452736ae52fd0c2a7ec44f27291a93ad18565e33ccdf31101a394L250-R270): Modified `build_libraries()` to handle the renaming and copying of the shared library for iOS builds, similar to the existing macOS logic.

### iOS Configuration:

* [`swig/config/ios-arm64-config.jam`](diffhunk://#diff-0c153443d3849916b104541f964a0f64ad99ae110017f83dc508ce7ef148079eR1-R9): Added a new configuration file for building with the `darwin` toolset targeting ARM64 architecture, specifying necessary compiler flags.
